### PR TITLE
Use new proposal protection format

### LINKED ISSUE #7196: Support local protection data interchange format
# 💎 Issue

### Background

in order to support validators to seamlessly migrate to and from prysm we have to support the [protection db data interchange
](https://hackmd.io/@sproul/Bk0Y0qdGD)
### Description
this is the list of steps to be taken:

- [x] create a new proposal schema update #7197
- [x] create a new attestation schema update #7248
- [x] support migration of data from old schema to new schema #7430
- [x] use the new schema for local protection #7518 
- [ ] validate in runtime
- [ ] support full yaml version data import
- [ ] support full yaml version data export
- [ ] add minimal protection method
- [ ] support minimal yaml version data import

### PR DESCRIPTION
**Pull request by shayzluf**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**
Use the new proposal protection format in runtime
**Why**
As part of the effort to support slashing protection data interchange format and the need to add signing root to our protection db i have found that proposals db was built for the old slashing rule (1 proposal per epoch) since that rule changed to (1 proposal per slot) there is no benefit for saving this data on epoch context and adding complexity .
The new db structure saves proposals signing roots per slot for slashing protection. 

- This PR also imports old protection data format to the new format on validator node startup and saves a flag of it being exported in the old bucket

**Which issues(s) does this PR fix?**

Part of #7196

data interchange format
https://hackmd.io/@sproul/Bk0Y0qdGD#Implementations

### CODE REVIEW

File: validator/client/propose.go [hunk 1]
```diff
@@ -200,20 +200,20 @@ func (v *validator) signRandaoReveal(ctx context.Context, pubKey [48]byte, epoch
 }
 
 // Sign block with proposer domain and private key.
-func (v *validator) signBlock(ctx context.Context, pubKey [48]byte, epoch uint64, b *ethpb.BeaconBlock) ([]byte, error) {
+func (v *validator) signBlock(ctx context.Context, pubKey [48]byte, epoch uint64, b *ethpb.BeaconBlock) ([]byte, *ethpb.DomainResponse, error) {
```
Comment (prestonvanloon) [inline validator/client/propose.go:203]: This method really should have a test.

File: validator/node/node.go [hunk 2]
```diff
@@ -104,6 +104,9 @@ func NewValidatorClient(cliCtx *cli.Context) (*ValidatorClient, error) {
 	if err := ValidatorClient.initializeFromCLI(cliCtx); err != nil {
 		return nil, err
 	}
+	//if err := ValidatorClient.db.UpdateProposalsProtectionDb(cliCtx.Context); err != nil {
+	//	return nil, err
+	//}
```
Comment (prestonvanloon) [inline validator/node/node.go:109]: Remove?

Comment (shayzluf) [reply validator/node/node.go]: was trying to find what causes the e2e failure

Comment (shayzluf) [reply validator/node/node.go]: removed the comment

File: validator/client/propose.go (hunk 1 above)
Comment (shayzluf) [reply validator/client/propose.go:203]: added static test for signature result from master

File: validator/db/kv/attestation_history_new.go [hunk 3]
```diff
@@ -198,3 +198,51 @@ func (store *Store) ImportOldAttestationFormat(ctx context.Context) error {
...
+				return err
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+
```
Comment (rauljordan) [inline validator/db/kv/attestation_history_new.go:229]:
```suggestion
	return err
```

File: validator/db/kv/attestation_history_new.go [hunk 4]
```diff
@@ -198,3 +198,51 @@ func (store *Store) ImportOldAttestationFormat(ctx context.Context) error {
 	err = store.SaveAttestationHistoryNewForPubKeys(ctx, dataMap)
 	return err
 }
+
+// UpdateAttestationProtectionDb exports old attestation protection data
+// format to the new format and save the exported flag to database.
+func (store *Store) UpdateAttestationProtectionDb(ctx context.Context) error {
+	importAttestations, err := store.shouldImportAttestations()
+	if err != nil {
+		return err
```
Comment (rauljordan) [inline validator/db/kv/attestation_history_new.go:207]: We should aim to use better formatted errors to diagnose failures at runtime, so perhaps errors.Wrap or fmt.Errorf with a better description for some of these

File: validator/db/kv/attestation_history_new.go [hunk 5]
```diff
@@ -198,3 +198,51 @@ func (store *Store) ImportOldAttestationFormat(ctx context.Context) error {
...
+	}
+	return nil
+
+}
+
+func (store *Store) shouldImportAttestations() (bool, error) {
+	var importAttestations bool
+	err := store.db.View(func(tx *bolt.Tx) error {
+
```
Comment (rauljordan) [inline validator/db/kv/attestation_history_new.go:235]:
```suggestion
```

File: validator/db/kv/new_proposal_history.go [hunk 6]
```diff
@@ -159,3 +160,56 @@ func pruneProposalHistoryBySlot(valBucket *bolt.Bucket, newestSlot uint64) error
...
+				return err
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (rauljordan) [inline validator/db/kv/new_proposal_history.go:191]:
```suggestion
	return err
```

File: validator/db/kv/new_proposal_history.go [hunk 7]
```diff
@@ -5,6 +5,7 @@ import (
 	"fmt"
 
 	"github.com/pkg/errors"
+	"github.com/prometheus/common/log"
```
Comment (rauljordan) [inline validator/db/kv/new_proposal_history.go:8]: We should be using logrus

**Review by terencechain — COMMENTED** [review]
@shayzluf where is the link for the latest format proposal?

We should check that aligns with this.

**Comment by shayzluf** [comment]
@terencechain added link to pr description

File: validator/db/kv/new_proposal_history.go (hunk 7 above)
Comment (shayzluf) [reply validator/db/kv/new_proposal_history.go]: removed

File: validator/db/kv/attestation_history_new.go (hunk 4 above)
Comment (shayzluf) [reply validator/db/kv/attestation_history_new.go]: added many erros.Wrap

File: validator/client/propose_protect.go [hunk 8]
```diff
@@ -29,7 +29,7 @@ func (v *validator) preBlockSignValidations(ctx context.Context, pubKey [48]byte
 		}
 
 		// If the bit for the current slot is marked, do not propose.
-		if slotBits.BitAt(block.Slot % params.BeaconConfig().SlotsPerEpoch) {
+		if !bytes.Equal(signingRoot, bytesutil.PadTo([]byte{}, 32)) {
```
Comment (nisdas) [inline validator/client/propose_protect.go:32]:
```suggestion
		if !bytes.Equal(signingRoot, params.BeaconConfig.ZeroHash[:]) {
```

File: validator/client/propose_protect.go [hunk 9]
```diff
@@ -74,15 +73,11 @@ func (v *validator) postBlockSignUpdate(ctx context.Context, pubKey [48]byte, bl
 	}
 
 	if featureconfig.Get().LocalProtection {
-		slotBits, err := v.db.ProposalHistoryForEpoch(ctx, pubKey[:], epoch)
+		signingRoot, err := helpers.ComputeSigningRoot(block.Block, domain.SignatureDomain)
 		if err != nil {
-			if v.emitAccountMetrics {
```
Comment (nisdas) [inline validator/client/propose_protect.go:79]: do we not want to check here ?

File: validator/db/kv/manage_test.go [hunk 10]
```diff
@@ -171,15 +171,15 @@ func prepareStore(store *Store, pubKeys [][48]byte) (*storeHistory, error) {
...
 
-func prepareStoreProposals(store *Store, pubKeys [][48]byte) (map[[48]byte]bitfield.Bitlist, error) {
-	proposals := make(map[[48]byte]bitfield.Bitlist)
+func prepareStoreProposals(store *Store, pubKeys [][48]byte) (map[[48]byte][]byte, error) {
+	proposals := make(map[[48]byte][]byte)
 
 	for i, key := range pubKeys {
-		proposalHistory := bitfield.Bitlist{byte(i), 0x00, 0x00, 0x00, 0x01}
-		if err := store.SaveProposalHistoryForEpoch(context.Background(), key[:], 0, proposalHistory); err != nil {
+		signingRoot := bytesutil.PadTo([]byte{byte(i)}, 32)
```
Comment (nisdas) [inline validator/db/kv/manage_test.go:178]: this is confusing, why is the signing root the index of the pubkey ?

Comment (shayzluf) [reply validator/db/kv/manage_test.go:178]: Just to have different values

File: validator/client/propose_protect.go (hunk 9 above)
Comment (shayzluf) [reply validator/client/propose_protect.go:79]: nice catch

**Review by rauljordan — DISMISSED** [review]

**Review by rauljordan — APPROVED** [review]

**Pull request merged**